README.md
# dobx_gen
Generate [dobx](https://github.com/dobx/dobx) observables via [source_gen](https://github.com/dart-lang/source_gen)

## Example
### lib/todo.dart
```
library hello;

import 'package:dobx/dobx.dart';
import 'package:dobx_gen/core.dart';

part 'hello.g.dart';

@dobx
abstract class Todo {

  String title;
  bool completed;

  factory Todo() => _$Todo( // this method is generated
    title: 'hello, world!',
    completed: false,
  );
}
```

### pubspec.yaml
```
name: hello
description: dobx_gen example
version: 0.1.0

dependencies:
  dobx: ^0.7.0
  dobx_gen: ^0.1.0

dev_dependencies:
  build_runner: ^0.3.0+1
```

### tool/build.dart
```
import 'dart:async';

import 'package:build_runner/build_runner.dart';
import 'package:dobx_gen/core_gen.dart';
import 'package:source_gen/source_gen.dart';

Future main(List<String> args) async {
  await build(
      new PhaseGroup.singleAction(
          new GeneratorBuilder([new DobxGenerator()]),
          new InputSet('hello', const ['lib/*.dart'])),
      deleteFilesByDefault: true);
}
```

### Generated code
```sh
dart tool/build.dart
```



CHANGELOG.md
## 0.2.0

* Ignore static fields

## 0.1.0

* Generate observable fields (string/int/bool) from a class annotated with @dobx


lib/core.dart
// ========================================================================
// Copyright 2017 David Yu
// ------------------------------------------------------------------------
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========================================================================
// @author dyu
// @created 2017/03/28

class Dobx {
  const Dobx();
}

const dobx = const Dobx();



e2e_example/lib/hello.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of hello;

// **************************************************************************
// Generator: DobxGenerator
// Target: abstract class Todo
// **************************************************************************

class _Todo extends PubSub implements Todo {
  String _title;
  String get title {
    $sub(1);
    return _title;
  }

  void set title(String title) {
    if (title != null && title == _title) return;
    _title = title ?? '';
    $pub(1);
  }

  bool _completed;
  bool get completed {
    $sub(2);
    return _completed;
  }

  void set completed(bool completed) {
    if (completed != null && completed == _completed) return;
    _completed = completed ?? false;
    $pub(2);
  }
}

Todo _$Todo({
  String title: '',
  bool completed: false,
}) {
  return new _Todo()
    .._title = title
    .._completed = completed;
}


lib/core_gen.dart
// ========================================================================
// Copyright 2017 David Yu
// ------------------------------------------------------------------------
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========================================================================
// @author dyu
// @created 2017/03/28

import 'dart:async';

import 'package:analyzer/dart/element/element.dart';
import 'package:source_gen/source_gen.dart';
import 'package:source_gen/src/utils.dart';

import 'package:build/build.dart';

import './core.dart';

class DobxGenerator extends GeneratorForAnnotation<Dobx> {
  const DobxGenerator();

  @override
  Future<String> generateForAnnotatedElement(
      Element element, Dobx annotation, BuildStep buildStep) async {
    if (element is! ClassElement) {
      var friendlyName = friendlyNameForElement(element);
      throw new InvalidGenerationSourceError(
          'Generator cannot target `$friendlyName`.',
          todo: 'Remove the @dobx annotation from `$friendlyName`.');
    }

    final ce = element as ClassElement;
    return _implClass(ce);
    //return annotation.mixin ? _subClass(ce) : _implClass(ce);
  }

  String _implClass(ClassElement ce) {
    var name = ce.name,
        body = new StringBuffer(),
        params = new StringBuffer(),
        assign_params = new StringBuffer(),
        i = 0;

    String ftype, fname, fval;
    for (var fe in ce.fields) {
      if (fe.isStatic)
        continue;

      ftype = fe.type.name;
      fname = fe.name;
      fval = _defaultValue(ftype);

      // the compiler allows '=' to assing default value but build_runner complains about it
      params.writeln('${ftype} ${fname} : $fval,');
      assign_params.writeln('.._${fname} = ${fname}');
      body.writeln('${ftype} _${fname};');
      body.writeln(_implMethod(++i, ftype, fname, fval));
    }

    return '''
    class _$name extends PubSub implements $name {
      $body
    }
    $name _\$$name({
      $params
    }) {
      return new _$name()
        $assign_params;
    }
    ''';
  }

  String _implMethod(int id, String type, String name, String val) {
    return '''
    $type get $name { \$sub($id); return _$name; }
    void set $name($type $name) {
      if ($name != null && $name == _$name) return;
      _$name = $name ?? $val;
      \$pub($id);
    }
    ''';
  }

  String _defaultValue(String type) {
    switch (type) {
      case 'String': return "''";
      case 'bool': return 'false';
      case 'double': return '0.0';
      case 'num': return '0';
      case 'int': return '0';
      default: return 'null';
    }
  }

  /*bool _isPrivateField(FieldElement fe) {
    // starts with '_'
    return 95 == fe.name.codeUnitAt(0);
  }

  String _subClass(ClassElement ce) {
    var name = ce.name,
        buf = new StringBuffer(),
        i = 0;

    String val;
    for (var fe in ce.fields) {
      if (_isPrivateField(fe) && 'null' != (val = _defaultValue(fe)))
        buf.writeln(_implMethod(++i, fe.type.name, fe.name.substring(1), val));
    }

    return '''
    class _$name extends $name with PubSub {
      _$name();
      $buf
    }
    $name _\$$name() { return new _$name(); }
    ''';
  }*/
}